File: leecode/code/394/394.py
class Solution:
    def decodeString(self, s: str) -> str:
        if len(s) == 0:
            return ""
        
        res = ""


        # 去掉字母
        i = 0
        while i < len(s):
            if (s[i] < '0' or s[i] > '9') and s[i] != '[' and s[i] != ']':
                i += 1
            else:
                break
        
        res = res + s[0: i]
        # 数字开头    
        s  = s[i:]
        i = 0
        num = 0
        while i < len(s) and s[i] <= '9' and s[i] >= '0':
            num = num * 10 + int(s[i])
            i += 1


        if num > 0:
            s = s[i:]
            i = 0
            count = 0
            if s[i] == '[':
                while i < len(s):
                    if s[i] == '[':
                        count += 1
                    elif s[i] == ']':
                        count -= 1
                    i += 1


                    if count == 0:
                        break
                
            temp = self.decodeString(s[1: i - 1])
            s = s[i:]
            if num > 0:
                for i in range(num):
                    res = res + temp


        return res + self.decodeString(s)

File: leecode/code/394/394.java
class Solution {


    public String dfs(String s) {
        if (s == "") {
            return "";
        }
        String res = "";
        int index = 0;
        while (index < s.length()) {
			if ((s.charAt(index) < '0' || s.charAt(index) > '9') && s.charAt(index) != '[' && s.charAt(index) != ']')
            res = res + s.charAt(index);
            index++;
        }
        String nums = "";
        while (index < s.length() && s.charAt(index) >= '0' && s.charAt(index) <= '9') {
            nums = nums + s.charAt(index);
            index++;
        }
        String temp = "";
        if (index < s.length() && s.charAt(index) == '[') {
            int count = 0;
            do {
                if (s.charAt(index) == '[') {
                    if (count != 0)
                        temp = temp + '[';
                    count++;
                } else if (s.charAt(index) == ']') {
                    count--;
                    if (count != 0)
                        temp = temp + ']';
                } else {
                    temp = temp + s.charAt(index);
                }
                index++;
            } while(index < s.length() && count != 0);
            temp = dfs(temp);
        }
        String k = "";
        if (nums.length() != 0) {
            int num = Integer.valueOf(nums);
            for (int i = 0; i < num; i++) {
                k = k + temp;
            }
        }
        String el = s.substring(index);
        return res + k + el;
    }



    public String decodeString(String s) {
        String res = s;
        while (res.indexOf('[') != -1) {
            res = dfs(res);
        }
        return res;
    }
}